api/cpp/include/slint.h

```diff
@@ -1049,6 +1049,182 @@ class MapModel : public Model<MappedModelData>
     std::function<MappedModelData(const SourceModelData &)> map_fn;
 };
 
+template<typename ModelData>
+class SortModel;
+
+namespace private_api {
+template<typename ModelData>
+struct SortModelInner : private_api::ModelChangeListener
+{
+    SortModelInner(std::shared_ptr<slint::Model<ModelData>> source_model,
+                   std::function<bool(const ModelData &, const ModelData &)> sort_fn,
+                   slint::SortModel<ModelData> &target_model)
+        : source_model(source_model), sort_fn(sort_fn), target_model(target_model)
+    {
+    }
+
+    void row_added(int first_inserted_row, int count) override
+    {
+        if (sorted_rows_dirty) {
+            reset();
+            return;
+        }
+
+        // Adjust the existing sorted row indices to match the updated source model
+        for (auto &row : sorted_rows) {
+            if (row >= first_inserted_row)
+                row += count;
+        }
+
+        for (int row = first_inserted_row; row < first_inserted_row + count; ++row) {
+
+            ModelData inserted_value = *source_model->row_data(row);
+            auto insertion_point =
+                    std::lower_bound(sorted_rows.begin(), sorted_rows.end(), inserted_value,
+                                     [this](int sorted_row, const ModelData &inserted_value) {
+                                         auto sorted_elem = source_model->row_data(sorted_row);
+                                         return sort_fn(*sorted_elem, inserted_value);
+                                     });
+
+            insertion_point = sorted_rows.insert(insertion_point, row);
+            target_model.row_added(std::distance(sorted_rows.begin(), insertion_point), 1);
+        }
+    }
+    void row_changed(int changed_row) override
+    {
+        if (sorted_rows_dirty) {
+            reset();
+            return;
+        }
+
+        auto removed_row_it =
+                sorted_rows.erase(std::find(sorted_rows.begin(), sorted_rows.end(), changed_row));
+        auto removed_row = std::distance(sorted_rows.begin(), removed_row_it);
+
+        ModelData changed_value = *source_model->row_data(changed_row);
+        auto insertion_point =
+                std::lower_bound(sorted_rows.begin(), sorted_rows.end(), changed_value,
+                                 [this](int sorted_row, const ModelData &changed_value) {
+                                     auto sorted_elem = source_model->row_data(sorted_row);
+                                     return sort_fn(*sorted_elem, changed_value);
+                                 });
+
+        insertion_point = sorted_rows.insert(insertion_point, changed_row);
+        auto inserted_row = std::distance(sorted_rows.begin(), insertion_point);
+
+        if (inserted_row == removed_row) {
+            target_model.row_changed(removed_row);
+        } else {
+            target_model.row_removed(removed_row, 1);
```

Comment:
when row_removed is called, the row was already added.  Meaning that the listener to row_removed that would query the model could get in an inconsistent state.
I think right now, nobody would query the model in row_removed,  but that may be a problem for later.

Reply:
Yeah, I think we have this in some other places as well - we need to define what exactly is allowed to be done in row_removed, i.e. what the expected state is. Right now I believe in none of the row_removed listeners we query data, but the moment we do this becomes an issue.